absl/random/internal/randen_detect.cc

```diff
@@ -45,6 +45,10 @@
 #if defined(ABSL_INTERNAL_USE_X86_CPUID)
 #if defined(_WIN32) || defined(_WIN64)
 #include <intrin.h>  // NOLINT(build/include_order)
+#elif ABSL_HAVE_BUILTIN(__cpuid)
+// MSVC-equivalent __cpuid intrinsic declaration for clang-like compilers
+// for non-Windows build environments.
+void __cpuid(int[4], int);
```

Comment:
Is the declaration required? Everywhere else where ABSL_HAVE_BUILTIN() returns true requires no declaration or #include. In other words, does the following work? I don't have this compiler so I can't check myself.

```
#elif !ABSL_HAVE_BUILTIN(__cpuid)
// MSVC-equivalent __cpuid intrinsic function.
static void __cpuid(int cpu_info[4], int info_type) {
  ...
}
```

Reply:
When I merge the two sections as you have suggested, I get this warning.

```
[build] [270/698  33% :: 7.426] Building CXX object absl/random/CMakeFiles/random_internal_randen_hwaes.dir/Debug/internal/randen_detect.cc.o
...
[build] .../abseil-cpp/absl/random/internal/randen_detect.cc:135:3: warning: implicitly declaring library function '__cpuid' with type 'void (int *, int) noexcept' [-Wimplicit-function-declaration]
[build]   __cpuid(reinterpret_cast<int*>(regs), 1);
[build]   ^
[build] .../abseil-cpp/absl/random/internal/randen_detect.cc:135:3: note: include the header <intrin.h> or explicitly provide a declaration for '__cpuid'
[build] 1 warning generated.
```

I do not see this same error with the normal clang++-16 compiler.

Reply:
Looking at the diffs of the implicit defines in the two compilers, I do not see anything that stands out.

```
echo | clang++-16 -E -dM - > /tmp/left
echo | icpx -E -dM - > /tmp/right
```

```
1a2
> #define __AMX_SUPPORTED__ 1
24a26
> #define __CLANG_CUDA_APPROX_TRANSCENDENTALS__ 1
41a44
> #define __FAST_MATH__ 1
115a119,120
> #define __INTEL_CLANG_COMPILER 20230100
> #define __INTEL_LLVM_COMPILER 20230100
168a174
> #define __I__ 1j
184a191
> #define __LONG_DOUBLE_SIZE__ 80
190c197
< #define __NO_INLINE__ 1
---
> #define __NO_MATH_ERRNO__ 1
197a205
> #define __OPTIMIZE__ 1
201,202d208
< #define __PIC__ 2
< #define __PIE__ 2
340c346
< #define __VERSION__ "Ubuntu Clang 16.0.0 (1~exp5ubuntu3)"
---
> #define __VERSION__ "Intel(R) oneAPI DPC++/C++ Compiler 2023.1.0 (2023.1.0.20230320)"
355c361
< #define __clang_version__ "16.0.0 (1~exp5ubuntu3)"
---
> #define __clang_version__ "16.0.0 (icx 2023.1.0.20230320)"
364,365d369
< #define __pic__ 2
< #define __pie__ 2
```

Reply:
Here is a simple repro.

/tmp/cpuid_test.cc:

```
#include <iostream>

static void __cpuid(int cpu_info[4], int info_type) {
  __asm__ volatile("cpuid \n\t"
                   : "=a"(cpu_info[0]), "=b"(cpu_info[1]), "=c"(cpu_info[2]),
                     "=d"(cpu_info[3])
                   : "a"(info_type), "c"(0));
}

int main() {
  int regs[4];
  __cpuid(reinterpret_cast<int*>(regs), 1);
  std::cout << "AES supported: " << ((regs[2] & (1 << 25)) == 0 ? "no" : "yes") << std::endl;  // AES
}
```

```
docker pull intel/oneapi-basekit:latest
```

```
docker run -it -v /tmp:/tmp intel/oneapi-basekit:latest icpx /tmp/cpuid_test.cc -o /tmp/cpuid_test2
```

```
/tmp/cpuid_test.cc:3:13: error: static declaration of '__cpuid' follows non-static declaration
static void __cpuid(int cpu_info[4], int info_type) {
            ^
/tmp/cpuid_test.cc:3:13: note: '__cpuid' is a builtin with type 2
1 error generated.
```